records/2023-01-26.bounce-rate-automated-emergency-brake.md

```diff
@@ -130,72 +134,82 @@ having
+The keys will expire within the chosen sliding window. Currently this window is defined as a 24 hour period.
 
-The new bouce rate calculation will be:
+The new bouce rate calculation for a service will be:
 
-$$Bounce Rate = \frac{\sum\limits_{n=1}^{24} \text{hard bounces n} +\text{length}( \text{hard bounces:\{service id\}} )}{\sum\limits_{n=1}^{24} \text{total notifications n} + \text{length}( \text{total notifications:\{service id\}} )}$$
+$$Bounce Rate = \frac {
+    \sum\limits_{n=1}^{24} \text{hard bounces seeded n}  +
+    \text{length} ( \text{sliding hard bounce} )
+}{
+    \sum\limits_{n=1}^{24}  \text{notifications seeded n}  +
+    \text{length} ( \text{sliding notifications} )
+}$$
 
 
 This above bounce rate is not an exact number but it an approximation of what the bounce rate is likely to be within a sliding window.
 
-#### Keys Task
+## Keys Task
 
 If we wanted to see all the keys for ALL the services - we should make an AWS ECS task that we can kick off.\
 Instead as an optimization, we should only seed the keys for a service where:
 1. A user from the service has logged into the ADMIN
 1. We get a message from the API for a certain service
 
 In both the above scenarios we will kick off a celery task that will see the bounce rate data. We need to ensure that if users of the service login, or we send multiple jobs through the API - we are not constantly starting the celery task to seed the bounce rate.
-In order for us to keep track, we will need to add a key to services table indicating if we have started to seed the bounce rate for that service. This will prevent duplication of jobs.
 
-#### Redis Memory Consideration
+## Redis Memory Consideration
+
+We are currently using 500mb on our Redis cluster which is assigned 1G of RAM. An increase in the number of keys means we might need to upgrade the cluster. [Current memory usage estimations and analysis](https://github.com/cds-snc/notification-attic/pull/31) and recent [Redis memory usage](https://gcdigital.slack.com/archives/CV38DBNVA/p1685372981086749) shows that, for the time being, this is not something we need to consider right now, but as Notify grows we will need to keep this in mind.
+
+# Additional considerations
+
+## Staggered roll-out
+To prevent suspending services that may not actually be causing issues with the overall bounce rate, it is suggested to roll out this feature in stages.
 
-We are currently using 500mb on our Redis cluster which is assigned 1G of RAM. An increase in the number of keys means we might need to upgrade the cluster.
+### Version 1
+The first stage would include tracking bounces, calculating bounce rates, and displaying them on service dashboards. It would also include providing bounce rate guidance for services.
 
-## Additional considerations
+### Version 1.5
+The second stages would include bug fixes, and any adjustments to the way we calculate bounce rate.
```

Comment:
```suggestion
This would include: bug fixes, displaying data from Redis in the dashboard, design changes to the dashboard to allow displaying 7 day statistics along side 24 hour bounce rate data, and any adjustments to the way we calculate bounce rate.
```